pkg/api/rest/update.go

```diff
@@ -37,9 +37,16 @@ type RESTUpdateStrategy interface {
 	// the object.  For example: remove fields that are not to be persisted,
 	// sort order-insensitive list fields, etc.
 	PrepareForUpdate(obj, old runtime.Object)
-	// ValidateUpdate is invoked after default fields in the object have been filled in before
-	// the object is persisted.
+	// ValidateUpdate is invoked after default fields in the object have
+	// been filled in, but before the object is persisted.  This method is
+	// responsible for ensuring that any disallowed changes are flagged.
+	// For example: changing immutable fields.  It does not need to
+	// re-validate the whole object - that is what Validate() is for.
 	ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList
+	// Validate is invoked after default fields in the object have been
+	// filled in and after the update itself has been validated, but before
+	// the object is persisted.
+	Validate(ctx api.Context, obj runtime.Object) fielderrors.ValidationErrorList
```

Comment:
I'd be fine with that.  Validate on strategy should then become ValidateCreate().  Update is distinct because of the possibility of change from a previous persisted version.

> On Mar 31, 2015, at 10:54 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/rest/update.go:
> 
> >  	ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList
> > +	// Validate is invoked after default fields in the object have been
> > +	// filled in and after the update itself has been validated, but before
> > +	// the object is persisted.
> > +	Validate(ctx api.Context, obj runtime.Object) fielderrors.ValidationErrorList
> It seems like we really want is to separate validation of create and update from validation of object invariants.
> 
> Create should validate that required fields are provided.
> 
> Update should validate that immutable fields are not changed.
> 
> Both should then invoke a shared function that validates requirements/invariants/rules of the object.
> 
> —
> Reply to this email directly or view it on GitHub.
>